extra/DEVELOPER_DOCS/StringStore-Vocab.md

```diff
@@ -0,0 +1,220 @@
+
+Besides `__getitem__`, the `StringStore` has functions to return specifically a
+string or specifically a hash, regardless of whether the input was a string or
+hash to begin with, though these are only used occasionally. (Maybe they can be
+removed?)
+
+### Implementation Details: Hashes and Allocations
+
+Hashes are 64-bit and are computed using murmurhash on UTF-8 bytes. There is no
+mechanism for detecting and avoiding collisions. To date there has never been a
+reproducible collision or user report about any related issues.
+
+The empty string is not hashed, it's just converted to/from 0. 
+
+A small number of strings use indices into a lookup table (so low integers)
+rather than hashes. This is mostly Universal Dependencies labels or other
+strings considered "core" in spaCy. This was critical in v1, which hasn't
+introduced hashing yet. Since v2 it's important for items in `spacy.attrs`,
+especially lexeme flags, but is otherwise only maintained for backwards
+compatability.
+
+You can call `strings["mystring"]` with a string the `StringStore` has never seen
+before and it will return a hash. But in order to do the reverse operation, you
+need to call `strings.add("mystring")` first. Without a call to `add` the
+string will not be interned.
+
+In normal use nothing is ever removed from the `StringStore`. In theory this
+means that if you do something like iterate through all hex values of a certain
+length you can have explosive memory usage. In practice this has never been an
+issue. (Note that this is also different from using `sys.intern` to intern
+Python strings, which does not guarantee they won't be memory collected later.)
+
+The way strings are stored in the `StringStore` is weird, but is well isolated
+from outside code. A short explanation is that very short strings are stored
+in a fixed-length buffer (one per string, not a shared buffer), while longer
+strings are prefixed with their length and use malloced memory. If you want to
+understand the full details see the implementation of `decode_Utf8Str` and
+`_allocate` in `strings.pyx`.
+
+### Notes on Python & Cython String Types
+
+The `StringStore` doesn't do anything too weird with Python strings, but because
+it's in Cython some of the type details of Python strings become important.
+
+In Python 2 strings were more like sequences of bytes, but in Python 3 they're
+more like UTF-8 sequences and treated separately from bytes. Cython has several
+string types to manage compatability with bytes and strings in Python 2 and 3:
+`str`, `unicode`, `bytes`, and `basestring`. 
+
+Here is how the types work in Python 3:
+
+- `str` and `unicode` are the same
+- `bytes` is just `bytes`
+- `basestring` is the union of `str` and `unicode`, so in Py3 all are equivalent
+
+`basestring` and `unicode` are still used in several places in spaCy code,
+presumably as a legacy of Python 2 support. In Python 3 there should be no
+difference between these types and `str`.
```

Comment:
This can be updated after the other PR is merged

Reply:
Since it's been merged I'll just remove this section.